### DISCUSSION TITLE
Passing variable to dynamic filter

### DISCUSSION THREAD

**Discussion by WojciechRhdt** [rappasoft/laravel-livewire-tables#1034] [Q&A] [answered]
Hello all,
I am trying to add a filter dynamically. This works fine so far. However, I need a variable from my foreach inside the filter function. Is there a way to pass this?

![filter](https://user-images.githubusercontent.com/5849082/215764008-f168266b-42dd-4afe-8371-4c5a930fd56a.PNG)

**Comment by lrljoe** [comment, marked as answer]
As this is an anonymous function, you'll need to pass the $criteria variable in, otherwise the function doesn't know that the variable exists.
e.g:
```
function (Builder $builder, float $value) use ($criteria) {
```
You will **definitely** have performance issues with your approach though, if you give some more detail around the use case, then I can probably point you in the right direction

**Reply by WojciechRhdt** [reply]
Thank you very much! My solution works now, but as you say there could be performance problems. This is my first Laravel project and I'm not that skilled yet.

I'm trying to build a table with information from multiple models. 

![papers](https://user-images.githubusercontent.com/5849082/216014703-ad50449e-3bd1-443f-b184-b82ac73c816a.PNG)


Papers
------
- id
- user_id
- title
- content

Reviews
-------
- id
- user_id
- paper_id
- status

Review_ratings
--------------
- id
- review_id
- criteria_id
- score

Criterias
---------
- id
- title
- weight

Each paper can have multiple reviews. Each review can evaluate the paper according to different criteria. Each criterion has a certain weight from which a score is calculated at the end (for each criterion and a total score). There is also another model for Topics. I would like to have all this information available in the table.

Not easy for me :) If you have any advice for me, I'd really appreciate it!

**Reply by lrljoe** [reply]
So you should definitely not be doing your joins in that fashion, as you'll end up loading the same tables multiple times pointlessly.

Instead, use the Builder function to create the initial query, using the "with", "withWhereHas" and "whereHas" functions with subqueries appropriately, in conjunction with the "when" function, so that you constrain the eager loads.

**[Constraining Eager Loads](https://laravel.com/docs/9.x/eloquent-relationships#constraining-eager-loads)**
```
use App\Models\Paper;

     /**
     * @return Builder
     */
    public function builder(): Builder
    {
      return Paper::with(['reviews' => function ($query) {
          $query->where('status', 1);
      }]);
    }
```

**[Querying Relationship Existence](https://laravel.com/docs/9.x/eloquent-relationships#querying-relationship-existence)**
```php
Paper::whereHas('reviews', function (Builder $query) {
    $query->where('content', 'like', 'code%');
})
```

Don't forget that you can use the "with" function to load sub-relations too, and you can constrain those equally, simple example below where:

1. "MyModel" has a relation with "phonenumbers"
2. "phonenumbers" has a relationship with "phonenumbertype"
3. I want to only retrieve (in this basic example) those phone numbers that the "name" field on the phonenumbertype model contains "Mobile"
e.g.
```php
    return MyModel::with(['phonenumbers' => function ($query) {
        $query->whereHas('phonenumbertype', function ($query) {
            $query->where('name', 'like', '%Mobile%');
        });
    }]);
```
You can obviously use whereIn instead for passing array values through if required, and use the "when" function to add in constraints from your filter.  You'll see less performance impact by constraining your eager loads, than by doing a where at the end!

Highly recommend you install [BarryVDH DebugBar](https://github.com/barryvdh/laravel-debugbar) and enable the Query Monitoring so that you can look at what queries are being generated.

It is also worth reviewing the hasManyThrough relation [See Here](https://laravel.com/docs/9.x/eloquent-relationships#has-many-through) within your Laravel Model as appropriate.

**Reply by WojciechRhdt** [reply]
Thank you very much for your help and time!

**Reply by lrljoe** [reply]
> Thank you very much for your help and time! 

No problem at all, if you do get stuck on something, or want a pointer, feel free to stick it in a public repo and @ me and I'll take a look when I get a moment.
